caliban_toolbox/reshape_data.py

```diff
@@ -0,0 +1,703 @@
+            slice_counter += 1
+
+    return slice_xr
+
+
+def create_slice_data(data_xr, slice_stack_len, slice_overlap=0):
+    """Takes an array of data and splits it up into smaller pieces along the stack dimension
+
+    Args:
+        data_xr: xarray of [fovs, stacks, crops, slices, rows, cols, channels] to be split up
+        slice_stack_len: number of z/t frames in each slice
+        slice_overlap: number of z/t frames in each slice that overlap one another
+
+    Returns:
+        xarray.DataArray: 7D tensor of sliced data
+        dict: relevant data for reconstructing original imaging after slicing
+    """
+
+    # sanitize inputs
+    if len(data_xr.shape) != 7:
+        raise ValueError('invalid input data shape, '
+                         'expected array of len(7), got {}'.format(data_xr.shape))
+
+    if slice_stack_len > data_xr.shape[1]:
+        raise ValueError('slice size is greater than stack length')
+
+    # compute indices for slices
+    stack_len = data_xr.shape[1]
+    slice_start_indices, slice_end_indices = \
+        compute_slice_indices(stack_len, slice_stack_len, slice_overlap)
+
+    slice_xr = slice_helper(data_xr, slice_start_indices, slice_end_indices)
+
+    log_data = {}
+    log_data['slice_start_indices'] = slice_start_indices.tolist()
+    log_data['slice_end_indices'] = slice_end_indices.tolist()
+    log_data['num_slices'] = len(slice_start_indices)
+
+    return slice_xr, log_data
+
+
+def save_npzs_for_caliban(resized_xr, original_xr, log_data, save_dir, blank_labels='include',
+                          save_format='npz', verbose=True):
+    """Take an array of processed image data and save as NPZ for caliban
+
+    Args:
+        resized_xr: 7D tensor of cropped and sliced data
+        original_xr: the unmodified xarray
+        log_data: data used to reconstruct images
+        save_dir: path to save the npz and JSON files
+        blank_labels: whether to include NPZs with blank labels (poor predictions)
+                      or skip (no cells)
+        save_format: format to save the data (currently only NPZ)
+        verbose: flag to control print statements
+    """
+
+    if not os.path.isdir(save_dir):
+        os.makedirs(save_dir)
+
+    # if these are present, it means data was cropped/sliced. Otherwise, default to 1
+    num_crops = log_data.get('num_crops', 1)
+    num_slices = log_data.get('num_slices', 1)
+
+    fov_names = original_xr.fovs.values
+    fov_len = len(fov_names)
+
+    if blank_labels not in ['skip', 'include', 'separate']:
+        raise ValueError('blank_labels must be one of '
+                         '[skip, include, separate], got {}'.format(blank_labels))
+
+    if blank_labels == 'separate':
+        os.makedirs(os.path.join(save_dir, 'separate'))
+
+    # loop through all crops in all images
+    for fov in range(fov_len):
+        for crop in range(num_crops):
+            for slice in range(num_slices):
```

Comment:
Glad we got it down to 3 from 4. Instead of using a triple for loop, I suggest looking into [`itertools.combinations`](https://docs.python.org/3/library/itertools.html#itertools.combinations). It is both easier to read and [apparently faster](https://stackoverflow.com/questions/37354372/itertools-vs-nested-loops-performance).

Reply:
It's so beautiful